### PR TITLE
[#22] Task queueing 

### PR DESCRIPTION
**Pull request by dkrawiec**

### CODE REVIEW

**Comment by dkrawiec** [comment]
Waiting for #46 to be merged to master

**Comment by dkrawiec** [comment]
@annadobrowolska 
Could you please review this today?

File: pom.xml [hunk 1]
```diff
@@ -31,12 +31,12 @@
...
+                <artifactId>eventbus-api</artifactId>
                 <version>${project.version}</version>
             </dependency>
             <dependency>
                 <groupId>kimage</groupId>
-                <artifactId>eventbus-api</artifactId>
+                <artifactId>eventbus-messageTranslator</artifactId>
                 <version>${project.version}</version>
             </dependency>
             <dependency>
```
Comment (annaskulimowska) [inline pom.xml:42]: ok

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageProcessingFlowTest.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ public void shouldFinishProcessing() throws InterruptedException {
         //Given
         ImageProcessingFlowFactory flowFactory = new ImageProcessingFlowFactory();
         Step step = new Step(0, PLUGIN_NAME, Arrays.asList(new StepDependency(0, Image.class)));
-        Task task = new Task("".getBytes(), Arrays.asList(step));
+        Task task = new Task("1", "".getBytes(), Arrays.asList(step));
         ImageProcessingFlow imageProcessingFlow = flowFactory.create(manifestRepository, eventBus, task);
         Image image = new Image();
         //When
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageProcessingFlowTest.java:72]: ok

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            if (res.succeeded()) {
+                Vertx vertx = res.result();
+                EventBus eventBus = vertx.eventBus();
+                eventBus.consumer(QUEUE_TOPIC).handler(objectMessage -> {
+                    Task task = messageTranslator.deserialize(Task.class, (String) objectMessage.body());
+                    taskQueue.add(task);
+                });
+
+            } else {
+                System.out.println("Unable to connect to the cluster");
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java:34]: System.out.println?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import io.vertx.core.eventbus.EventBus;
+import pl.edu.uj.kimage.api.Result;
+import pl.edu.uj.kimage.eventbus.JsonMessageTranslator;
+
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class OutgoingTaskHandler implements Runnable {
+    private static Vertx vertx;
+    private Queue<CalculationResult> taskQueue = new LinkedBlockingQueue<>();
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:14]: final

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    private JsonMessageTranslator messageTranslator = new JsonMessageTranslator();
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:15]: final

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java (hunk 3 above)
Comment (dkrawiec) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java:34]: No logger...

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    private volatile boolean running = true;
+
+    public OutgoingTaskHandler() {
+        VertxOptions options = new VertxOptions();
+        Vertx.clusteredVertx(options, res -> {
+            if (res.succeeded()) {
+                vertx = res.result();
+            } else {
+                System.out.println("Unable to connect to cluster");
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:25]: System.out.println

Comment (dkrawiec) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:25]: No logger...

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java (hunk 3 above)
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java:34]: ok, but you should left TODO

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java (hunk 6 above)
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:25]: so, where is TODO?

Comment (dkrawiec) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:25]: In pending commit

Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:25]: ok

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java [hunk 7]
```diff
@@ -45,6 +45,7 @@ public void run() {
         }
     }
 
+
     public void addFinishedTask(CalculationResult result) {
         taskQueue.add(result);
     }
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:54]: ok

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java [hunk 8]
```diff
@@ -13,16 +13,19 @@
...
     public IncomingTaskHandler() {
         VertxOptions options = new VertxOptions();
         Vertx.clusteredVertx(options, res -> {
             if (res.succeeded()) {
-                Vertx vertx = res.result();
+                vertx = res.result();
                 EventBus eventBus = vertx.eventBus();
                 eventBus.consumer(QUEUE_TOPIC).handler(objectMessage -> {
                     Task task = messageTranslator.deserialize(Task.class, (String) objectMessage.body());
+                    System.out.println("Got task " + task.getTaskId());
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java:27]: System.out.println

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java [hunk 9]
```diff
@@ -13,16 +13,19 @@
...
                     taskQueue.add(task);
+                    objectMessage.reply("GOT IT");
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/IncomingTaskHandler.java]: got what? maybe add id or something?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java [hunk 10]
```diff
@@ -35,6 +35,7 @@ public void run() {
                 Object result = calculationResult.getResult();
                 String className = calculationResult.getResultClass().getName();
                 Result resultJson = new Result(className, messageTranslator.serialize(result).getBytes());
+                System.out.println("Publishing result for task " + calculationResult.getTaskId());
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/queueing/OutgoingTaskHandler.java:40]: lack of TODO

File: imageprocessor/imageprocessor-plugin/pom.xml [hunk 11]
```diff
@@ -20,6 +20,15 @@
...
+            <artifactId>assertj-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
     </dependencies>
 
 </project>
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-plugin/pom.xml:34]: ok

**Comment by annaskulimowska** [comment]
I ended my review

**Comment by dkrawiec** [comment]
Please give me a minute and you could do a re-review. I need this changes soon.

**Comment by annaskulimowska** [comment]
ok

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/queueing/IncomingTaskHandlerTest.java [hunk 12]
```diff
@@ -49,7 +49,7 @@ public void shouldAddReceivedTaskToQueue() throws Exception {
 
 
         //Then
-        assertThat(future.get()).isEqualTo("GOT IT");
+        assertThat(future.get()).isEqualTo("GOT " + taskId);
         assertThat(incomingTaskHandler.takeTask().getTaskId()).isEqualTo(taskId);
     }
 }
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/queueing/IncomingTaskHandlerTest.java:55]: ok

**Comment by annaskulimowska** [comment]
ok

**Pull request merged**